### PR TITLE
coord: refactor connection cleanup and cancellation

### PR DESCRIPTION
**Pull request by mjibson**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5621)

### CODE REVIEW

**Comment by mjibson** [comment]
I may have gone crazy with the `FetchResult` stuff but I think it looks pretty good?

File: src/pgwire/src/protocol.rs [hunk 1]
```diff
@@ -139,6 +147,13 @@ where
             None => "eof",
         };
 
+        // Clear any cancellation message.
+        // TODO(mjibson): This makes the use of .changed annoying since it will
+        // generally always have a NotCancelled message first that needs to be ignored,
+        // and thus run in a loop. Figure out a way to have the future only resolve on
+        // a Cancelled message.
+        let _ = self.cancel_tx.lock().unwrap().send(Cancelled::NotCancelled);
```
Comment (benesch) [inline src/pgwire/src/protocol.rs:155]: If you're on board with moving this to `SessionClient` (see below), then I think there is a really nice place for this logic:

```rust
impl SessionClient {
    async fn canceled(&self) {
        loop {
            self.cancel_rx.changed().await;
            if let Cancelled::Cancelled = *cancel.borrow() {
                return;
            }
        }
    }
}
```

File: test/lang/csharp/test.sh [hunk 2]
```diff
@@ -19,5 +19,4 @@ cd "$(dirname "$0")/../../.."
 
 cd test/lang/csharp
 
-dotnet test csharp-npgsql4.csproj
-dotnet test csharp-npgsql5.csproj
+dotnet test csharp-npgsql.csproj
```
Comment (benesch) [inline test/lang/csharp/test.sh:22]: Are we dropping support for Npgsql 4?

**Review by benesch — COMMENTED** [review]
Awesome! The coordinator/session/transaction side of this looks great to me. I scattered a proposal within about the protocol side though—I think stashing away the guts of the cancelation implementation inside the `coord::SessionClient` could make for a nice clean API!

File: test/lang/csharp/test.sh (hunk 2 above)
Comment (mjibson) [reply test/lang/csharp/test.sh:22]: See the (updated) commit message: npgsql4 still works, but the type of the exception changed and I didn't want to deal with splitting this up.

**Comment by mjibson** [comment]
Thanks for the very useful comments. Implemented all your suggestions.

**Review by benesch — APPROVED** [review]
Looks so good!

**Pull request merged**